zerver/lib/narrow.py

```diff
@@ -919,6 +927,30 @@ def get_channel_from_narrow_access_unchecked(
     return None
 
 
+def update_narrow_terms_containing_with_operator(
+    realm: Realm, narrow: OptionalNarrowListT
+) -> OptionalNarrowListT:
+    if narrow is not None:
+        contains_with_operator = list(filter(lambda term: term["operator"] == "with", narrow))
+
+        if len(contains_with_operator) > 0:
+            with_term = contains_with_operator[0]
+            message = get_recipient_topic_by_narrow_operand_access_unchecked(
+                realm, int(with_term["operand"])
+            )
+            if message is not None and message.is_stream_message():
+                for term in narrow:
+                    if term["operator"] == "channel":
+                        stream_id = message.recipient.type_id
+                        term["operand"] = get_stream_by_id_in_realm(stream_id, realm).name
+                    elif term["operator"] == "topic":
+                        term["operand"] = message.topic_name()
+
+            narrow.remove(with_term)
+
+    return narrow
```

Comment:
I think we will want this function to also return at least a `boolean` for whether the narrow was in fact modified from its original state; I think there's a good chance we'll want to have the API response do something to reflect whether that in fact occurred.

Reply:
we can decide it in the client side whether the narrow has been updated or not.